### ISSUE TITLE
Discussion: make variables generalised in the type in scope for the clauses?

### ISSUE THREAD

**Issue by gallais** [agda/agda#3447]
An artificial example: I think I would like to have `variablelift` behave the same
as `modulelift`. If a variable has been generalised in a type then the name should
be available when defining the clauses.

This way when there are multiple generalised variables, the user doesn't have to
guess in which order to write the various `{a = a}` that they need to be able to
mention them.

```agda
open import Agda.Primitive using (Level; _⊔_)

record Lift a ℓ (A : Set a) : Set (a ⊔ ℓ) where
  field lower : A

module _ {a} where

  sectionlift : ∀ ℓ → Set a → Set (a ⊔ ℓ)
  sectionlift = Lift a

variable
  a : Level

variablelift : ∀ ℓ → Set a → Set (a ⊔ ℓ)
variablelift = Lift a

-- Generalizable variable genmodule.a is not supported here
-- when scope checking a
```

In practice I have data which can be inferred from some of the function's arguments
and thus can be left implicit but sometimes needs to be mentioned in the RHS of a
clause or in a `with`).

**Comment by jespercockx** [comment]
Isn't this the same as https://github.com/agda/agda/issues/3111?

**Comment by UlfNorell** [comment]
No, that issue suggests introducing implicit foralls in the right-hand side and this issue is about making variables generalised in the type be in scope.
```agda
foo : T n → Set
foo t = T n

foo-issue3447 : T n → Set
foo-issue3447 {n = n} t = T n

foo-issue3111 : T n → Set
foo-issue3111 t = ∀ {n} → T n
```

**Comment by nad** [comment]
Am I right in thinking that we can close at least one of these two issues?

**Comment by UlfNorell** [comment]
Indeed.
